File: VisitorPattern/VisitorPattern.cs
using System;
using System.Collections.Generic;

namespace VisitorPattern
{
    public interface IDocumentVisitor
    {
        void Visit(TextFile document);
        void Visit(Spreadsheet document);
        void Visit(PresentationFile document);
    }

    public interface IDocument
    {
        void Accept(IDocumentVisitor visitor);
    }

    public class TextFile : IDocument
    {
        public void Accept(IDocumentVisitor visitor) => visitor.Visit(this);
    }

    public class Spreadsheet : IDocument
    {
        public void Accept(IDocumentVisitor visitor) => visitor.Visit(this);
    }

    public class PresentationFile : IDocument
    {
        public void Accept(IDocumentVisitor visitor) => visitor.Visit(this);
    }

    public class WordCounter : IDocumentVisitor
    {
        public void Visit(TextFile document) => Console.WriteLine("WordCounter -> TextFile");
        public void Visit(Spreadsheet document) => Console.WriteLine("WordCounter -> Spreadsheet");
        public void Visit(PresentationFile document) => Console.WriteLine("WordCounter -> PresentationFile");
    }

    public class TextExtractor : IDocumentVisitor
    {
        public void Visit(TextFile document) => Console.WriteLine("TextExtractor -> TextFile");
        public void Visit(Spreadsheet document) => Console.WriteLine("TextExtractor -> Spreadsheet");
        public void Visit(PresentationFile document) => Console.WriteLine("TextExtractor -> PresentationFile");
    }

    public class FormatAnalyzer : IDocumentVisitor
    {
        public void Visit(TextFile document) => Console.WriteLine("FormatAnalyzer -> TextFile");
        public void Visit(Spreadsheet document) => Console.WriteLine("FormatAnalyzer -> Spreadsheet");
        public void Visit(PresentationFile document) => Console.WriteLine("FormatAnalyzer -> PresentationFile");
    }

    class Program
    {
        static void Main()
        {
            List<IDocument> documents = new List<IDocument>
            {
                new TextFile(),
                new Spreadsheet(),
                new PresentationFile()
            };

            IDocumentVisitor wordCounter = new WordCounter();
            IDocumentVisitor textExtractor = new TextExtractor();
            IDocumentVisitor formatAnalyzer = new FormatAnalyzer();

            foreach (var doc in documents) doc.Accept(wordCounter);
            foreach (var doc in documents) doc.Accept(textExtractor);
            foreach (var doc in documents) doc.Accept(formatAnalyzer);
        }
    }
}

File: VisitorPattern/VisitorPattern.cpp
#include <iostream>
#include <vector>

class TextFile;
class Spreadsheet;
class PresentationFile;

class IDocumentVisitor {
public:
    virtual ~IDocumentVisitor() = default;

    virtual void visit(TextFile* document) = 0;
    virtual void visit(Spreadsheet* document) = 0;
    virtual void visit(PresentationFile* document) = 0;
};

class IDocument {
public:
    virtual ~IDocument() = default;
    virtual void accept(IDocumentVisitor* visitor) = 0;
};

class TextFile : public IDocument {
public:
    
    void accept(IDocumentVisitor* visitor) override;
};

class Spreadsheet : public IDocument {
public:
    
    void accept(IDocumentVisitor* visitor) override;
};

class PresentationFile : public IDocument {
public:
    
    void accept(IDocumentVisitor* visitor) override;
};

class WordCounter : public IDocumentVisitor {
public:

    void visit(TextFile* document) override {
        std::cout << "WordCounter -> TextFile\n";
    }
    void visit(Spreadsheet* document) override {
        std::cout << "WordCounter -> Spreadsheet\n";
    }
    void visit(PresentationFile* document) override {
        std::cout << "WordCounter -> PresentationFile\n";
    }
};

class TextExtractor : public IDocumentVisitor {
public:

    void visit(TextFile* document) override {
        std::cout << "TextExtractor -> TextFile\n";
    }
    void visit(Spreadsheet* document) override {
        std::cout << "TextExtractor -> Spreadsheet\n";
    }
    void visit(PresentationFile* document) override {
        std::cout << "TextExtractor -> PresentationFile\n";
    }
};

class FormatAnalyzer : public IDocumentVisitor {
public:

    void visit(TextFile* document) override {
        std::cout << "FormatAnalyzer -> TextFile\n";
    }
    void visit(Spreadsheet* document) override {
        std::cout << "FormatAnalyzer -> Spreadsheet\n";
    }
    void visit(PresentationFile* document) override {
        std::cout << "FormatAnalyzer -> PresentationFile\n";
    }
};

void TextFile::accept(IDocumentVisitor* visitor) {
    visitor->visit(this);
}

void Spreadsheet::accept(IDocumentVisitor* visitor) {
    visitor->visit(this);
}

void PresentationFile::accept(IDocumentVisitor* visitor) {
    visitor->visit(this);
}

int main() {
    std::vector<IDocument*> documents = {
        new TextFile(),
        new Spreadsheet(),
        new PresentationFile()
    };

    WordCounter wordCounter;
    TextExtractor textExtractor;
    FormatAnalyzer formatAnalyzer;

    for (auto doc : documents) {
        doc->accept(&wordCounter);
    }

    for (auto doc : documents) {
        doc->accept(&textExtractor);
    }

    for (auto doc : documents) {
        doc->accept(&formatAnalyzer);
    }

    for (auto doc : documents) {
        delete doc;
    }

    return 0;
}
